File: hadoop-native-core/rpc/call.c
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "rpc/call.h"
#include "rpc/proxy.h"

#include <stdlib.h>
#include <string.h>
#include <uv.h>

void hrpc_call_deliver_err(struct hrpc_call *call, struct hadoop_err *err)
{
    hrpc_raw_cb_t cb = call->cb;
    void *cb_data = call->cb_data;
    free(call->payload.base);
    memset(call, 0, sizeof(*call));
    __sync_fetch_and_or(&call->active, 0);
    cb(NULL, err, cb_data);
}

void hrpc_call_deliver_resp(struct hrpc_call *call, struct hrpc_response *resp)
{
    hrpc_raw_cb_t cb = call->cb;
    void *cb_data = call->cb_data;
    free(call->payload.base);
    memset(call, 0, sizeof(*call));
    __sync_fetch_and_or(&call->active, 0);
    cb(resp, NULL, cb_data);
}

int hrpc_call_activate(struct hrpc_call *call)
{
    return !!__sync_bool_compare_and_swap(&call->active, 0, 1);
}

void hrpc_call_deactivate(struct hrpc_call *call)
{
    __sync_fetch_and_and(&call->active, 0);
}

int hrpc_call_is_active(const struct hrpc_call *call)
{
    return !!__sync_fetch_and_or((int*)&call->active, 0);
}

// vim: ts=4:sw=4:tw=79:et

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/ipc/Call.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.hadoop.hdfs.ipc;

import com.google.protobuf.Message;
import com.google.protobuf.RpcCallback;
import java.io.IOException;
import org.apache.hadoop.classification.InterfaceAudience;

@InterfaceAudience.Private
class Call {
  private final int id;

  private final String protocolName;

  private final long protocolVersion;

  private final String methodName;

  private final Message param;

  private final Message responseDefaultType;

  private volatile Message response;

  private volatile IOException error;

  private boolean done;

  private final RpcCallback<Call> callback;

  Call(int id, String protocolName, long protocolVersion, String methodName,
      Message param, Message responseDefaultType, RpcCallback<Call> callback) {
    this.id = id;
    this.protocolName = protocolName;
    this.protocolVersion = protocolVersion;
    this.methodName = methodName;
    this.param = param;
    this.responseDefaultType = responseDefaultType;
    this.callback = callback;
  }

  private void callComplete() {
    callback.run(this);
  }

  /**
   * Set the exception when there is an error. Notify the caller the call is
   * done.
   *
   * @param error exception thrown by the call; either local or remote
   */
  void setException(IOException error) {
    synchronized (this) {
      if (done) {
        return;
      }
      this.done = true;
      this.error = error;
    }
    callComplete();
  }

  /**
   * Set the return value when there is no error. Notify the caller the call is
   * done.
   *
   * @param response return value of the call.
   * @param cells Can be null
   */
  void setResponse(Message response) {
    synchronized (this) {
      if (done) {
        return;
      }
      this.done = true;
      this.response = response;
    }
    callComplete();
  }

  int getId() {
    return id;
  }

  String getProtocolName() {
    return protocolName;
  }

  long getProtocolVersion() {
    return protocolVersion;
  }

  String getMethodName() {
    return methodName;
  }

  Message getParam() {
    return param;
  }

  Message getResponseDefaultType() {
    return responseDefaultType;
  }

  Message getResponse() {
    return response;
  }

  IOException getError() {
    return error;
  }

  boolean isDone() {
    return done;
  }
}